fix: markdown rendering

hackice20 opened the pull request:
- Fixes #24454 
- Fixes CAL-6573

## What does this PR do?

Fixes markdown rendering on public booking pages and in email notifications. Previously, markdown (headers, lists, bold, italic, links, etc.) was displayed as raw text instead of formatted HTML.

## Technical Changes:

Implemented custom markdown parser in markdownToSafeHTML.ts (server-side) and markdownToSafeHTMLClient.ts (client-side)
Added descriptionAsSafeHTML prop to event type queries (getPublicEvent.ts, getServerSideProps.tsx)
Updated CalEventParser.ts to render markdown in email descriptions using markdownToSafeHTML
Security: removed unsafe un-escaping of URLs, added double-quote escaping, fixed double-escaping in code blocks
Fixed ordered list numbering by using value attribute on list elements
Configured DOMPurify and sanitize-html to allow style attributes for inline CSS (email compatibility)

## Why This Matters:

Users often copy event descriptions from GPT/LLMs, which output markdown. Without rendering, headers, lists, and formatting appear as raw syntax, reducing readability. This fix ensures markdown is properly rendered on booking pages and in emails.

## Visual Demo (For contributors especially)
#### Video Demo (if applicable):

https://github.com/user-attachments/assets/a6cea20d-1c26-4ce2-a879-9e3ab91eb952





- Show screen recordings of the issue or feature.
- Demonstrate how to reproduce the issue, the behavior before and after the change.

#### Image Demo (if applicable):

- Add side-by-side screenshots of the original and updated change.
- Highlight any significant change(s).

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

### Environment Setup:
- No special environment variables required

- Ensure dev server is running: yarn dev
- Test Data:
- Create or edit an event type with a description containing:
- Headers: # H1, ## H2, ### H3
- Lists: - item or 1. item
- Bold: **text**
- Italic: *text* or _text_ or __text__
- Links: [text](https://example.com)

### Test Steps:

- Edit event type at /event-types/[id]?tabName=setup
- Add markdown in the description field
- Save and view the public booking page: /[username]/[event-slug]
- Verify markdown renders as formatted HTML
- Create a test booking and check the confirmation email
- Verify markdown renders correctly in the email HTML
- Expected Behavior:
- Markdown syntax is converted to styled HTML
- Headers display with appropriate font sizes
- Lists show proper bullets/numbers
- Links are clickable and open in new tabs
- Bold/italic text is formatted correctly
- No XSS vulnerabilities (malicious scripts are sanitized)

### Security Testing:

- Test with malicious markdown: [click](javascript:alert('xss')) - should be sanitized
- Test with HTML injection attempts - should be escaped

## Checklist

- [x] I have read the contributing guide
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings

<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/calcom/cal.com/pull/25052" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

hackice20 committed:
fix: render markdown in booking pages and emails

hackice20 committed:
fix: improve security and fix escaping bugs in markdown parser

hackice20 committed:
Fix markdown rendering and code style issues

hackice20 committed:
feat: Improve markdown list rendering for consecutive blank lines

hackice20 committed:
fix: Prevent closing lists for list items at indent 0

- Check if line is a list item before closing lists in indent-0 block
- Fixes issue where top-level lists were incorrectly terminated
- Consecutive list items now stay in the same list block

dhairyashiil reviewed (changes requested):
Hello @hackice20, Thanks a lot for the work on this PR, it is clear you put a lot of effort into it

However, to keep things consistent and maintainable in the codebase, could we please stick with our existing approach using the `markdown-it` library plus sanitization, instead of introducing a custom parser? 
We already standardize on `markdown-it` (see `@calcom/lib/markdownIt`).

Could you also follow the simpler approach used in PR #24677 and check the review points I left there to ensure those are covered here as well

_PS: If there is a specific capability you need that `markdown-it` doesn't cover, please call it out, we can look at adjusting the config_

hackice20 committed:
refactor: replace custom markdown parser with markdown-it

- Replace 300+ line custom parser with standardized markdown-it library
- Add inline styles via renderer rules for email compatibility
- Fix bullet point rendering with list-style-type: disc
- Fix italic rendering (_text_) by unescaping underscores
- Fix bold text rendering by improving list normalization regex
- Fix infinite recursion in renderer rules
- Reduce code by 53% (364 lines removed)
- Maintain backward compatibility (no breaking changes)

Files changed:
- packages/lib/markdownIt.ts: Enhanced with renderer rules for inline styles
- packages/lib/markdownToSafeHTML.ts: Replaced custom parser (369→105 lines)
- packages/lib/markdownToSafeHTMLClient.ts: Replaced custom parser (297→77 lines)
- packages/lib/CalEventParser.ts: Minor type fix

Fixes:
- Bullet points now render correctly with list-style-type: disc
- Italics (_text_) now work correctly
- Bold (**text**) no longer shows stray asterisks
- All markdown features work: headers, lists, links, code, etc.

hackice20 committed:
fix: address cubic bot review comments - make markdown stripping and unescaping more conservative

- stripMarkdown: Only remove actual markdown syntax patterns, not all punctuation (fixes 'Dial #1234' corruption)
- unescapeMarkdown: Only unescape when legacy Turndown data is detected (preserves intentional escapes like \*literal\*)

hackice20 committed:
hotfix

hackice20 committed:
Update packages/lib/markdownIt.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

dhairyashiil reviewed (changes requested):
please address these comments and add unit tests for XSS prevention and markdown edge cases

hackice20 committed:
feat: improve markdown sanitization security and add tests

dhairyashiil reviewed (changes requested):
Looks good Yash, Great work!

On the last step, since it’s a very big code change, I don’t want this to go into production without thorough testing. This way, we avoid reverting the PR in the future. 

Here are some points to test:

- Create a booking with markdown in the description.
- Accept the meeting in Outlook.
- Verify that the description shows clean text (no ****).
- Check the .ics file content directly.
- Test HTML email rendering.
- Test plain text email rendering.
- Test with special characters.

Please attach a video of the results. 🙏🏼

hackice20 committed:
fix: remove hardcoded markdown text colors

hackice20 committed:
Update packages/lib/markdownIt.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

hackice20 committed:
addressing cubic bot

hackice20 committed:
ignoring cubic bot

hackice20 commented:
Hello @dhairyashiil ,any updates?

dhairyashiil commented:
Thank you for the PR @hackice20. I’m not entirely sure which of the points mentioned above by cubic we should address and which we can ignore. I’ll ask a team member to take a look. Thanks!

hackice20 commented:
> Thank you for the PR @hackice20. I’m not entirely sure which of the points mentioned above by cubic we should address and which we can ignore. I’ll ask a team member to take a look. Thanks!

Yes Sure, No worries. Thanks !

hackice20 committed:
fix: resolve merge conflict in CalEventParser.ts

hackice20 commented:
Hello @anikdhabal please review this PR, it looks big but the tests are huge after confirmation i can shrink the PR size.
